examples/ct/ct_handlers_test.go

```diff
@@ -222,13 +255,195 @@ func TestGetRoots(t *testing.T) {
+	km := setupMockKeyManager(toSign)
+	client := new(trillian.MockTrillianLogClient)
+
+	roots := loadCertsIntoPoolOrDie(t, []string{testonly.FakeCACertPem})
+	reqHandlers := CTRequestHandlers{0x42, roots, client, km, time.Millisecond * 500, fakeTimeSource}
+
+	pool := loadCertsIntoPoolOrDie(t, []string{testonly.LeafSignedByFakeIntermediateCertPem, testonly.FakeIntermediateCertPem})
+	chain := createJsonChain(t, *pool)
+
+	leaves := leafProtosForCert(t, km, pool.RawCertificates())
+
+	client.On("QueueLeaves", mock.MatchedBy(deadlineMatcher), &trillian.QueueLeavesRequest{LogId: 0x42, Leaves: leaves}, mock.Anything /* []grpc.CallOption */).Return(&trillian.QueueLeavesResponse{Status: &trillian.TrillianApiStatus{StatusCode: trillian.TrillianApiStatusCode_OK}}, nil)
+
+	recorder := makeAddChainRequest(t, reqHandlers, chain)
+
+	assert.Equal(t, http.StatusOK, recorder.Code, "expected HTTP OK for valid add-chain: %v", recorder.Body)
+	km.AssertExpectations(t)
+	client.AssertExpectations(t)
+
+	// Roundtrip the response and make sure it's sensible
+	var resp addChainResponse
+	err := json.NewDecoder(recorder.Body).Decode(&resp)
+
+	assert.NoError(t, err, "failed to unmarshal json: %v", recorder.Body.Bytes())
+
+	assert.Equal(t, ct.V1, ct.Version(resp.SctVersion))
+	assert.Equal(t, ctMockLogID, resp.ID)
+	assert.Equal(t, uint64(1469185273000000), resp.Timestamp)
+	assert.Equal(t, "BAEABnNpZ25lZA==", resp.Signature)
+}
+
+func loadCertsIntoPoolOrDie(t *testing.T, certs []string) *PEMCertPool {
+	pool := NewPEMCertPool()
+
+	for _, cert := range certs {
+		ok := pool.AppendCertsFromPEM([]byte(cert))
+
+		if !ok {
+			t.Fatalf("couldn't parse test certs: %v", certs)
+		}
 	}
 
 	return pool
 }
+
+func createJsonChain(t *testing.T, p PEMCertPool) io.Reader {
+	var chain jsonChain
+
+	for _, rawCert := range p.RawCertificates() {
+		b64 := base64.StdEncoding.EncodeToString(rawCert.Raw)
+		chain.Chain = append(chain.Chain, b64)
+	}
+
+	var buffer bytes.Buffer
+	// It's tempting to avoid creating and flushing the intermediate writer but it doesn't work
+	writer := bufio.NewWriter(&buffer)
+	err := json.NewEncoder(writer).Encode(&chain)
+	writer.Flush()
+
+	if err != nil {
+		t.Fatalf("failed to create test json: %v", chain)
+	}
+
+	return bufio.NewReader(&buffer)
+}
+
+func leafProtosForCert(t *testing.T, km crypto.KeyManager, certs []*x509.Certificate) []*trillian.LeafProto {
+	// We don't care about the SCT as that's sent to the client and we're testing frontend ->
+	// backend interaction
+	merkleLeaf, _, err := SignV1SCTForCertificate(km, certs[0], fakeTime)
+
+	if err != nil {
+		t.Fatalf("failed to sign test SCT or Merkle Leaf: %v", err)
+	}
+
+	var b bytes.Buffer
+	if err := WriteMerkleTreeLeaf(&b, merkleLeaf); err != nil {
+		t.Fatalf("failed to serialize leaf: %v", err)
+	}
+
+	leafHash := sha256.Sum256(b.Bytes())
```

Comment:
Could do with a comment saying that it's not an RFC6962 leaf hash, to avoid confusion.